### PR TITLE
UI-20 add number field

### PR DESCRIPTION
**Pull request by lmiller1990**
Basically copypaste of TextInput but changed validations a bit. For now only supporting positiiveOrWholeNumbers for this ticket: https://doseme.atlassian.net/browse/AA-79

Will add more validations as we good

image

### CODE REVIEW

File: src/NumberInput/validation.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+const validate = (rules: IValidators, text: string): IValidationResult => {
+  if (rules.isRequired && !text.length) {
+    return {
+      message: 'This field is required',
+      valid: false,
+    }
+  }
+
+  // 0 or greater
+  if (rules.onlyPositiveWholeNumber && !isNormalInteger(text)) {
```
Comment (MichaelDorkhom) [inline src/NumberInput/validation.ts]: Aren't all whole numbers positive? Depending on how flexible we want to be I'd either shorten this prop's name, or make it a `type` that can be things like ['whole', 'integer', ..]

File: src/NumberInput/validation.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  return {
+    valid: true,
+  }
+}
+
+export {
+  validate,
+}
```
Comment (MichaelDorkhom) [inline src/NumberInput/validation.ts]: Next line

File: src/NumberInput/validation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  test('onlyPositiveWholeNumber valid', () => {
+    const actual = validate(rules, '123')
+    expect(actual).toEqual({ valid: true })
+  })
+
+  test('onlyPositiveWholeNumber valid', () => {
+    const actual = validate(rules, '0')
+    expect(actual).toEqual({ valid: true })
+  })
+})
```
Comment (MichaelDorkhom) [inline src/NumberInput/validation.spec.ts]: Next line

File: src/NumberInput/validation.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  message?: string
+}
+
+// https://stackoverflow.com/questions/10834796/validate-that-a-string-is-a-positive-integer/10834843
+const isNormalInteger = (str: string): boolean => {
+  const n = Math.floor(Number(str))
+  return n !== Infinity && String(n) === str && n >= 0
+}
+
+const validate = (rules: IValidators, text: string): IValidationResult => {
```
Comment (MichaelDorkhom) [inline src/NumberInput/validation.ts]: Can we make this a reusable function? Things like `isRequired` should be shared across the different input fields

File: src/NumberInput/NumberInput.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    defaultValue,
+    ...rest
+  } = props
+
+  const [error, setError] = useState<string>('')
+  const [isValid, setValid] = useState<boolean>(true)
+
+  const handleValidate = (str: string): boolean => {
+    const status = validate({
+      onlyPositiveWholeNumber
```
Comment (MichaelDorkhom) [inline src/NumberInput/NumberInput.tsx]: Don't we also want to validate `isRequired`?

File: src/NumberInput/NumberInput.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import React from 'react'
+import noop from 'lodash/noop'
+import { storiesOf } from '@storybook/react'
+
+import { NumberInput } from './NumberInput'
+
+
```
Comment (MichaelDorkhom) [inline src/NumberInput/NumberInput.stories.tsx]: Double blank line?

File: src/NumberInput/validation.spec.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  validate,
+  IValidators,
+  IValidationResult,
+} from './validation'
+
+describe('validate', () => {
+  const rules: IValidators = {
+    onlyPositiveWholeNumber: true,
+    isRequired: true
+
```
Comment (MichaelDorkhom) [inline src/NumberInput/validation.spec.ts]: Blank line. We'll need to take some time and review lint rules so these can be caught

File: src/NumberInput/validation.ts (hunk 1 above)
Comment (lmiller1990) [reply src/NumberInput/validation.ts]: -10 a whole number that no positive. a whole number is number without decimals. like `int`.

File: src/NumberInput/validation.spec.ts (hunk 7 above)
Comment (lmiller1990) [reply src/NumberInput/validation.spec.ts]: yes. definitely - PRs should not be about formatting

**Pull request merged**